dexlib2/src/main/java/org/jf/dexlib2/builder/MethodLocation.java

```diff
@@ -74,167 +73,28 @@ public int getIndex() {
         return index;
     }
 
-    @Nonnull
-    private List<Label> getMutableLabels() {
-        if (labels == null) {
-            labels = new ArrayList<>(1);
-        }
-        return labels;
-    }
-    
-    @Nonnull
-    private List<Label> getImmutableLabels() {
-        if (labels == null) {
-            return ImmutableList.of();
-        }
-        return labels;
-    }
-
-
-    @Nonnull
-    private List<BuilderDebugItem> getMutableDebugItems() {
-        if (debugItems == null) {
-            debugItems = new ArrayList<>(1);
-        }
-        return debugItems;
-    }
-
-    @Nonnull
-    private List<BuilderDebugItem> getImmutableDebugItems() {
-        if (debugItems == null) {
-            return ImmutableList.of();
-        }
-        return debugItems;
-    }
-
-    @Nonnull
-    private List<BuilderDebugItem> getDebugItems(boolean mutable) {
-        if (debugItems == null) {
-            if (mutable) {
-                debugItems = new ArrayList<>(1);
-                return debugItems;
-            }
-            return ImmutableList.of();
-        }
-        return debugItems;
-    }
-
     void mergeInto(@Nonnull MethodLocation other) {
-        if (this.labels != null || other.labels != null) {
-            List<Label> otherLabels = other.getMutableLabels();
-            for (Label label: this.getImmutableLabels()) {
-                label.location = other;
-                otherLabels.add(label);
-            }
-            this.labels = null;
-        }
+        labels.mergeItemsInto(other, other.labels);
 
-        if (this.debugItems != null || other.labels != null) {
-            // We need to keep the debug items in the same order. We add the other debug items to this list, then reassign
-            // the list.
-            List<BuilderDebugItem> debugItems = getMutableDebugItems();
-            for (BuilderDebugItem debugItem: debugItems) {
-                debugItem.location = other;
-            }
-            debugItems.addAll(other.getImmutableDebugItems());
-            other.debugItems = debugItems;
-            this.debugItems = null;
-        }
+        debugItems.mergeItemsInto(other, other.debugItems);
```

Comment:
The labels and debug items have slightly different merge logic in the original version. I couldn't remember exactly why, but looking at where this is used, the only usage is when deleting an instruction, and merging the existing debug items/labels from the deleted instruction into the next instruction.

For labels, it doesn't matter what order they are in, since any label at a particular location is equivalent to any other label.

But for debug items, the order does matter. It looks like the original method was written assuming that "other" is the next instruction, although I didn't remember this, and it's not immediately apparent.

Can you rename the helper method "mergeItemsIntoNext", rename the parameter from "other" to "next", and then make sure that the items from "this" get placed into the list before the items from "next", so that they maintain the same order relative to one another.

Reply:
Yes about renaming.

Regarding the order - changed it as you required.

It makes me suspect that the function `getDebugItems` should not return a `set`.
https://github.com/JesusFreke/smali/blob/7079014a29869e1bac22226681de3471b6dc11b5/dexlib2/src/main/java/org/jf/dexlib2/builder/MethodLocation.java#L176

It implements iteration over a set by iterating over a list.

What do you think ?

Reply:
Yeah, I agree. Since the order can be important, a list makes more sense.